Test TensorTypeSet instead of autograd_meta_ for variable-ness.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28620 Merge Tensor and Variable.
* #28610 Null AutogradMeta optimization
* #28609 Delete dead TensorImpl::detach_autograd_meta
* #28602 Remove unused gradient_edge argument from make_variable_view
* #28593 Add an AutogradMeta factory.
* #28601 Unfriend Variable factory functions.
* #28592 Move all autograd_meta_ manipulating operations out-of-line.
* **#28543 Test TensorTypeSet instead of autograd_meta_ for variable-ness.**

By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171157](https://our.internmc.facebook.com/intern/diff/D18171157)

ezyang committed:
Test TensorTypeSet instead of autograd_meta_ for variable-ness.

By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

yf225 commented on c10/core/TensorImpl.h:
```diff
@@ -1579,7 +1580,13 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  // of std::default_delete<AutogradMeta> which won't work as AutogradMeta
+  // is incomplete at this point.  But the defaulting to nullptr is also
+  // pointless because unique_ptr has a default constructor which will do the
+  // right thing.
```
Is the `NB:` part of the comment meant for another PR? Because it says `We CANNOT default this to nullptr` but we still do in this PR :D

ezyang replied:
Yeah it's wrong. I just fixed it

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171157](https://our.internmc.facebook.com/intern/diff/D18171157)

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171157](https://our.internmc.facebook.com/intern/diff/D18171157)

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171157](https://our.internmc.facebook.com/intern/diff/D18171157)

[ghstack-poisoned]

ezyang committed:
Update on "Test TensorTypeSet instead of autograd_meta_ for variable-ness."


By the current autograd_meta_ <=> type_set_ invariant (now explicitly documented
in the right place!), these are equivalent.  But when I introduce null
autograd_meta_ optimization, they won't be equivalent anymore: TensorTypeSet is
going to give me the right information no matter what.

In the long run, this patch will be a wash, because everything will "be a variable"
in the long term.  But I am making this change now to make sure that the invariant
actually holds.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171157](https://our.internmc.facebook.com/intern/diff/D18171157)

[ghstack-poisoned]
